---
title: "Saldana, Montano, et al. -- 09 patients_bulkrnaseq developmental signatures"
author: "Luis Montano"
date: '2023-11-14'
output: html_document
params:
  inputdata: '~/path/to/input_data/'
  plotpath: '~/path/to/plots/'
  cachedir: '~/path/to/cachedir/'
  outpath: '~/mnt_out/figures/'
  resourcepath: "~/mnt_resources/"
  gtfpath: '~/path/to/gencode.v40.annotation.gtf.gz'
  downloadedfilepath: '~/path/to/downloaded_files/'
  fig_width: 6
  fig_height: 4
---

# Analysis of developmental gene signatures and their association with patient survival

## Dependencies

Output from script scrna/03, specifically the WT dataset and the cellinfo data structure for the WT markers. 
Most of the inputs come from script scrna/06_projecting_patient_cells_to_NCNB_mutantclusters.Rmd but have been added hear for ease. 

## Data requirements

This script depends on data obtained from three different sources:

a)  SEQC neuroblastoma dataset
b)  TARGET NB
c)  CCRI

To run the following scripts, follow these instructions:

1.  Create a folder named bulkrna_patients inside params\$inputdata. All patient data shall be placed herein.

2.  Obtain the data in the following ways:

```{=html}
<!-- -->
```
a)  SEQC:

```{=html}
<!-- -->
```
1.  Download and unzip the raw counts: <https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE49711&format=file&file=GSE49711%5FSEQC%5FNB%5FMAV%5FG%5Flog2%2E20121127%2Etxt%2Egz>
2.  Download the clinical covariates: clinical covariates: from bioprotocols publication <https://bio-protocol.org/exchange/minidetail?type=30&id=9552520>
3. Download additional metadata from https://ftp.ncbi.nlm.nih.gov/geo/series/GSE62nnn/GSE62564/matrix/

```{=html}
<!-- -->
```
b)  TARGET: manual) go to <https://portal.gdc.cancer.gov/analysis_page?app=CohortBuilder> and download the STAR counts for the TARGET-NBL open cohort (130 samples). A merged table for this data can be provided upon request. automatic) follow the targetgdc chunk below for import.

c)  CCRI: Data can be provided upon request, subject to terms of use.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
source("~/R/scutils.R")
source("~/R/importdata.R")
library(DESeq2)
library(GSVA)
library(survival)
library(survminer)
library(TCGAbiolinks)
library(ggfortify)
library(ggExtra)
library(gridExtra)
library(ggforce)
library(ggrastr)

options(future.globals.maxSize = 6 * 10^9)
vfeatures=5000
pcdims=30
umapdims=10
sz=.02
threed=F #use if a 3d umap is desired
tsca=200


  numcells=100

################################################################################
# recreation/ reloading controls
################################################################################  
  fullrecreate.list=list(
  cors=T,
  rgeneposint=T,
  fulltable=T
  )
    
fullreload.list=list(
  cors=F,
  rgeneposint=F,
  fulltable=F)

  fullrecreate=function(x){

  if(!(x %in% names(fullrecreate.list))){
    return(F)
    }else{return(fullrecreate.list[[x]])}
}

    fullreload=function(x){

  if(!(x %in% names(fullreload.list))){
    return(T)
    }else{return(fullreload.list[[x]])}
}  
  

################################################################################
#glasswork plot parameters
################################################################################
gwvarr="stage"

  sca=10
  rr=600
  ssca=200
  varrlab=gwvarr
  sca=10
  rr=600
  ssca=180
  hullpointsize=0.1
  fillvar="clusterfrac"
colorpoints="type"
plotmeth="jitter"
concavity=2

################################################################################
#  specific functions
################################################################################

format.status=Vectorize(function(x) {
  if(!is.na(x)){if(x=="Alive" || x=="0"){0}else{
  if(x=="Dead" || x=="1"){1}}}else{NA}
  }, USE.NAMES=F)

edit.stage23= Vectorize(function(st){
 if(grepl("Stage 2", st, ignore.case=T) | grepl("Stage 3", st, ignore.case=T)){
  return("Stage 2-3") 
 }else{
   if(grepl("Stage 4s", st, ignore.case=T)){
   return("Stage 4s")
     }else{
return(st)}}}, USE.NAMES=F) 


```

load marker lists

```{r}

####  markers 
#that are present in patient cells triaged into the respective the mutant reference clusters . these do not have to be shared with the actual invitro clusters, they just pop up in the patient cells.

markers.pts.mut= list(M1 = c("PNMA5", "IER3", "ID1", "AC023509.4", "AL031714.1", 
"FOSB", "KLF4", "JUN", "GADD45G", "AC044849.1", "AC093635.1", 
"AC097103.2", "ID4", "ID3", "RBM4", "TOB1", "FOS", "IER2", "TRIB1", 
"JUNB"), M10 = c("IGLC2", "IFITM1", "IFIT1", "S100A10", "SAMD9", 
"ISG15", "ISLR", "GADD45B", "GBP2", "RGS1", "APOL6", "MX1", "ISG20", 
"CASP4", "XAF1", "TNFAIP6", "IFI27", "HES1", "IFITM3"), M11 = c("SNHG7", 
"MLF1", "SNHG15", "ZNF830", "FGFR1OP2", "GOLGA7", "PSMC6"), M12 = c("MT-CO3", 
"MTRNR2L12", "MT-CO1", "MT-CYB", "MT-CO2", "MT-ND5", "MT-ATP6", 
"MT-ND1", "MT-ND4", "MT-ND4L", "MT-ND2", "SF1", "MALAT1", "CCNL1", 
"RNMT"), M13 = c("AUTS2", "ANK2", "CADM1", "CHGB"), M15 = c("CAVIN1", 
"TAGLN", "MYL9", "ACTA2", "KRT18", "CTGF", "THBS1", "SULF1", 
"NNMT", "CCDC80", "PRRX1", "PHLDA2", "ANXA1", "CAVIN3", "COL1A1", 
"ANGPTL4"), M16 = c("FOXS1", "GJA4", "TINAGL1", "SOD3", "BST2", 
"CCND2", "KIAA0040", "MIR4435-2HG", "SELENOP", "IFITM2", "CAV1", 
"CYTOR", "RARRES3", "IGFBP7", "MGP", "VAMP5"), M4 = c("CENPN", 
"CDK1", "UBE2C", "MRPL21", "KPNA2", "RANBP1"),M5mm = c("HIST1H1D", "HIST1H1B", "CPLX3", "NPW", "HIST1H1C", "HPDL", "CENPM", "CDT1", "PSMG1", "DSCC1", 
"XRCC2", "TRIP13", "SAC3D1", "RAD51AP1", "RRP9", "UNG", "CDCA4", 
"EXOSC5", "MAD2L1", "POLR3K", "DCTPP1", "TYMS", "CENPW", "MND1", 
"CKS1B", "DUT"),  M5 = c("HIST1H1D", 
"HIST1H1B", "CPLX3", "NPW", "HIST1H1C", "MYCNOS", "HPDL", "CENPM", 
"MYCN", "CDT1", "PSMG1", "DSCC1", "XRCC2", "TRIP13", "SAC3D1", 
"RAD51AP1", "RRP9", "UNG", "CDCA4", "EXOSC5"), M8 = c("PPP1R17", 
"PPP1R14A", "SSTR2", "NYAP2", "GKAP1", "PAPSS1", "CD47"))

#obtained from cellinfo.pts.shared, markers that are shared by wt clusters and 
#patient cells triaged into those clusters
markers.pts.wt.shared=list(C11 = c("TAGLN", "CAVIN1", "MYL9", "ACTA2", "IGFBP7", "KRT18", 
"RRAS", "CYTOR", "CCDC80", "MIR4435-2HG", "PPIC", "EVA1B", "ANXA1", 
"BGN", "COL1A1", "MSRB3", "MYLK", "SPARC", "COL4A1", "PHLDA2", 
"COL5A1", "COL1A2", "COL3A1", "S100A11", "COL5A2", "POSTN", "FN1", 
"PRSS23", "TNFRSF12A", "COL6A3", "TIMP1", "IGFBP6", "COL6A2", 
"COL4A2", "PDGFRB", "FSTL1", "LOXL2", "FBN1", "TPM2", "TGFBI", 
"TPM1", "MYH9", "CALD1", "TUBB6", "TGFB1I1", "EFEMP2", "ITGA1", 
"AHNAK", "IER3", "CDH11", "H2AFJ", "PDLIM3", "ITGAV", "ARID5B", 
"ACTN1", "PALLD", "EMILIN1", "TAGLN2", "DSTN"), C13 = c("DLC1", 
"RBFOX1", "RORA", "RIMS2", "AUTS2", "SLC8A1", "UNC5C", "DPP6", 
"ANK2"), C14 = c("LINC00682", "PHOX2A"), C2 = "PRKDC", 
    C4 = "GKAP1", C5 = c("SIX1", "PPP1R17"), C9 = "IGFBP5")


################################################################################
#importing from Rmd 06
################################################################################
#markers.mutant=cellinfo.ptsm2$markerlist
#markers.wt.shared=cellinfo.pts.shared$markerlist

#plugging them here for ease.

markers.mutant=
list(M1 = c("PNMA5", "IER3", "ID1", "AC023509.4", "AL031714.1", 
"FOSB", "KLF4", "JUN", "GADD45G", "AC044849.1", "AC093635.1", 
"AC097103.2", "ID4", "ID3", "RBM4", "TOB1", "FOS", "IER2", "TRIB1", 
"JUNB"), M10 = c("IGLC2", "IFITM1", "IFIT1", "S100A10", "SAMD9", 
"ISG15", "ISLR", "GADD45B", "GBP2", "RGS1", "APOL6", "MX1", "ISG20", 
"CASP4", "XAF1", "TNFAIP6", "IFI27", "HES1", "IFITM3"), M11 = c("SNHG7", 
"MLF1", "SNHG15", "ZNF830", "FGFR1OP2", "GOLGA7", "PSMC6"), M13 = c("AUTS2", 
"ANK2", "CADM1", "CHGB", "ICA1"), M15 = c("CAVIN1", "TAGLN", 
"MYL9", "ACTA2", "KRT18", "CTGF", "THBS1", "SULF1", "NNMT", "CCDC80", 
"PRRX1", "ANXA1", "PHLDA2", "CAVIN3", "COL1A1", "ANGPTL4"), M16 = c("FOXS1", 
"GJA4", "TINAGL1", "SOD3", "BST2", "CCND2", "KIAA0040", "SELENOP", 
"MIR4435-2HG", "IFITM2", "RARRES3", "CYTOR", "CAV1", "IGFBP7", 
"MGP", "VAMP5"), M4 = c("CENPN", "CDK1", "UBE2C", "MRPL21", "KPNA2", 
"RANBP1"), M5 = c("HIST1H1D", "HIST1H1B", "CPLX3", "NPW", "HIST1H1C", 
"MYCNOS", "HPDL", "CENPM", "MYCN", "CDT1", "PSMG1", "XRCC2", 
"DSCC1", "TRIP13", "SAC3D1", "RAD51AP1", "RRP9", "UNG", "CDCA4", 
"EXOSC5"), M8 = c("PPP1R17", "PPP1R14A", "SSTR2", "NYAP2", "GKAP1", 
"PAPSS1"))

markers.wt.shared=
list(C11 = c("TAGLN", "CAVIN1", "MYL9", "ACTA2", "IGFBP7", "KRT18", 
"RRAS", "CYTOR", "CCDC80", "MIR4435-2HG", "PPIC", "EVA1B", "ANXA1", 
"BGN", "COL1A1", "MSRB3", "MYLK", "SPARC", "COL4A1", "LGALS1", 
"PHLDA2", "COL5A1", "COL1A2", "COL3A1", "S100A11", "COL5A2", 
"FN1", "PRSS23", "TNFRSF12A", "COL6A3", "TIMP1", "IGFBP6", "COL6A2", 
"COL4A2", "PDGFRB", "FSTL1", "LOXL2", "FBN1", "TPM2", "TGFBI", 
"TPM1", "MYH9", "CALD1", "TUBB6", "TGFB1I1", "EFEMP2", "ITGA1", 
"AHNAK", "IER3", "CDH11", "H2AFJ", "PDLIM3", "ITGAV", "ARID5B", 
"ACTN1", "PALLD", "EMILIN1", "TAGLN2", "DSTN"), C13 = c("DLC1", 
"RBFOX1", "RORA", "CTNNA2", "RIMS2", "AUTS2", "SLC8A1", "UNC5C", 
"RBMS3", "SYT1"), C14 = c("LINC00682", "PHOX2A"), C4 = "GKAP1", C5 = c("SIX1", "PPP1R17"
    ), C9 = "IGFBP5")

markers.mut.shared=list(
    M13 = "CHGB", M15 = c("TAGLN", "ACTA2"),
    M5 = c("NPW", "RRP9"), M8 = "PPP1R17")

```

Load target dataset (gdc)

```{r targetgdc}

query <- GDCquery(
    project = "TARGET-NBL",
    data.category = c("Transcriptome Profiling"),
    data.type = "Gene Expression Quantification", 
    workflow.type = "STAR - Counts")

# remove duplicated cases
query$results[[1]]=query$results[[1]][!duplicated(query$results[[1]]$cases),]

GDCdownload(query = query)

counts.target.online <- GDCprepare(
    query = query,
    save = TRUE, 
    save.filename = file.path("~/mnt_out/ncnb2/input_data/bulkrna_patients/gdcquery.rds")
    )

metadata.target.online <- GDCquery_clinic("TARGET-NBL", "clinical")


```

Load TARGET data (manual)

```{r targetmanual}


target.data.path=file.path(params$datapath, "bulkrna_patients/target/")
metadata.target0= fread(file.path(params$datapath, "bulkrna_patients/target_metadata/20231127_TARGET_clinical_data_merged.csv")) %>% as.data.frame %>% col2names(., "File.ID") %>% mutate(sample.id=File.ID)

metadata.target.cnas0=fread(file.path(params$datapath, "bulkrna_patients/target_metadata/target_clinical_covariates_pugh_sup_table_1.csv")) %>% as.data.frame 
metadata.target.cnas1=metadata.target.cnas0[, 1:30] %>% fixnames %>% mutate(Case.ID=Case.USI)

emptycnas=(metadata.target.cnas1 %>% pull(Case.ID))==""
metadata.target.cnas= metadata.target.cnas1[!emptycnas,]

chr.annotations.list=lapply(metadata.target0 %>% pull(Case.ID) %>% unique, function(x) metadata.target.cnas %>% filter(Case.ID==x)) 

chr.annotations=chr.annotations.list[lapply(chr.annotations.list, function(x) nrow(x)) %>% Reduce(c,.) %>% as.logical %>% which] %>% Reduce(rbind, .)


metadata.target=metadata.target0 %>% left_join(., chr.annotations, by="Case.ID") %>% mutate(x17q_gain=X17q., x1q_gain=X1q.) %>% col2names(., "File.ID")

allfiles.intable=metadata.target %>% rownames

allfiles.indirectory=list.files(target.data.path)

allfiles.indirectory= allfiles.indirectory[!grepl("MANIFEST", allfiles.indirectory)]

##  files shared between both metadata and files
shared.files=intersect(allfiles.intable, allfiles.indirectory)

count.column="unstranded"
alltablist=lapply(shared.files, function(ptfile){
 
pt.target.counts=list.files(file.path(target.data.path, ptfile), pattern="*augmented_star_gene_counts.tsv")   

table.header=fread(file.path(target.data.path, ptfile, pt.target.counts), nrow=2) %>% colnames 
table.main0=fread(file.path(target.data.path, ptfile, pt.target.counts), skip=6) %>% as.data.frame  
## remove duplicated entries and get the right column
table.main=table.main0[!(table.main0$V2 %>% duplicated),] %>% col2names(., "V2")

colnames(table.main)=table.header

table.main=table.main %>% select(!!sym(count.column)) %>% givecolnames(., 1, ptfile)
table.main
}) 

############################################################################
# finding synonyms for marker genes and harmonising them across datasets
############################################################################

counts.target = alltablist %>% Reduce(cbind, .)
synonym.h2afj=c( "H2AJ", "FLJ10903","MGC921")[c( "H2AJ", "FLJ10903","MGC921") %in% rownames(counts.target)]
rownames(counts.target)[rownames(counts.target)==synonym.h2afj]="H2AFJ"

fwrite(counts.target, file=file.path(params$datapath, "bulkrna_patients/target_counts.csv"))

################################################################################
# FILTER PATIENTS FOR USE DURING ANALYSIS
################################################################################


target.before.filtering=transferplot(so=NULL, df=metadata.target, labelvars = c("primary_diagnosis", "inss_stage", "MYCN_status", "Sample.Type", "tissue_or_organ_of_origin") , colorlist=allcolors)+ggtitle("Overview of entire dataset")

edit.event.target=Vectorize(function(ev){

  if(ev=="Censored"){
    return(0)
  }else{return(1)
 }},USE.NAMES=F)


edit.event.target.2=Vectorize(function(efstime, ostime, firstevent){
  ##Conditions for 1: 
  #a) if first event is death and efstime == ostime
  #b) if first event is not death and if efstime is shorter than ostime 

  if((firstevent=="Death" & efstime==ostime)  || (firstevent!="Death" & efstime < ostime)){
  return(1)}else{
    return(0)
  } }, USE.NAMES=F)
  

fwrite(metadata.target, file=file.path(params$datapath, "bulkrna_patients/target_processed/metadata_target_ready4import.csv"), row.names=T)  

metadata.target2=fread(file=file.path(params$datapath, "bulkrna_patients/target_processed/metadata_target_ready4import.csv")) %>% as.data.frame %>% col2names(., "File.ID")


mt=metadata.target2 %>% filter(primary_diagnosis=="Neuroblastoma, NOS") %>% mutate(Event_Free_Survival=edit.event.target.2(Event_Free_Survival_Time_in_Days, Overall_Survival_Time_in_Days, First_Event), inss_stage=edit.stage23(inss_stage))

  
fwrite(mt, file=file.path(params$datapath, "bulkrna_patients/target_processed/metadata_target_filtered_ready4analysis.csv"))  

mt=fread( file=file.path(params$datapath, "bulkrna_patients/target_processed/metadata_target_filtered_ready4analysis.csv")) %>% as.data.frame %>%  col2names(., "File.ID")

counts.target.filtered=counts.target[, rownames(mt)]

fwrite(counts.target.filtered, file=file.path(params$datapath, "bulkrna_patients/target_processed/target_counts_filtered_onlyNB_unstranded.csv")) 

target.after.filtering=transferplot(so=NULL, df=mt, labelvars = c("primary_diagnosis", "inss_stage", "MYCN_status", "Sample.Type", "tissue_or_organ_of_origin") , colorlist=allcolors)+ggtitle("Overview of entire dataset")

```

load seqc dataset

```{r loadseqc}

format.seqc.entry=function(x) if(grepl(":", x)){strsplit(x, split="\\: ")[[1]][2]}else{x} 
get.seqc.field=function(x) if(grepl(":", x)){strsplit(x, split="\\: ")[[1]][1]}else{x} 
clean.name=Vectorize(function(x) {
  first.split=strsplit(x, split=" \\[")[[1]][1]
  
  if(grepl("SEQC", first.split) ){
  strsplit(first.split, split="SEQC_")[[1]][2]}else{first.split}}, USE.NAMES=F)

remove.first.row=function(df)
  df[2:nrow(df), ]
remove.last.row=function(df)
  df[1:(nrow(df)-1), ]

##seqc counts

counts.seqc0=fread(file.path(params$datapath, "bulkrna_patients/seqc/GSE49711_SEQC_NB_MAV_G_log2.20121127.txt")) %>% as.data.frame

counts.seqc=counts.seqc0[,c(1, 10:(ncol(counts.seqc0)))] 
newnames=clean.name(colnames(counts.seqc))
newnames[1]="gene"             
counts.seqc=  counts.seqc %>% givecolnames(., nms=newnames) 
counts.seqc=(counts.seqc[!duplicated(counts.seqc[, "gene"]), ] %>% col2names(., "gene"))[, 2:ncol(counts.seqc)]

############################################################################
# finding synonyms for marker genes and harmonising them across datasets
############################################################################

synonym.ppp1r17=c("GSBS", "C7orf16", "PPP1R17")[c("GSBS", "C7orf16", "PPP1R17") %in% rownames(counts.seqc)]
synonym.LINC=c("NONHSAG037830.2", "HSALNG0034072", "LINC00682")[c("NONHSAG037830.2", "HSALNG0034072", "LINC00682") %in% rownames(counts.seqc)] 
synonym.rbfox1= c("HRNBP1","A2BP1","FOX-1","A2BP","2BP1")[ c("HRNBP1","A2BP1","FOX-1","A2BP","2BP1") %in% rownames(counts.seqc)] 
synonym.cavin1=c("CGL4","PTRF","FKSG13","CAVIN")[c("CGL4","PTRF","FKSG13","CAVIN") %in% rownames(counts.seqc)] 
synonym.cytor= c("C2orf59","C2ORF59" ,"NCRNA00152","LINC00152","MGC4677" )[c("C2orf59","C2ORF59" ,"NCRNA00152","LINC00152","MGC4677" ) %in% rownames(counts.seqc)] 
synonym.eva1b=c("FLJ10647","C1orf78","C1ORF78","FAM176B")[c("FLJ10647","C1orf78","C1ORF78","FAM176B") %in% rownames(counts.seqc)] 
synonym.mir=c("LINC00978","MORRBID","AGD2","AK001796","LncRNA-AWPPH","MIR4435-1HG","NONHSAG028949.2","NONHSAG028940.2","HSALNG0017865","LNCRNA-AWPPH","Lnc-ANAPC1-4"
,"Lnc-ANAPC1-7"
,"MIR4435-2HG")[c("LINC00978","MORRBID","AGD2","AK001796","LncRNA-AWPPH","MIR4435-1HG","NONHSAG028949.2","NONHSAG028940.2","HSALNG0017865","LNCRNA-AWPPH","Lnc-ANAPC1-4"
,"Lnc-ANAPC1-7"
,"MIR4435-2HG") %in% rownames(counts.seqc)] 


rownames(counts.seqc)[rownames(counts.seqc)==synonym.ppp1r17]="PPP1R17"
rownames(counts.seqc)[rownames(counts.seqc)==synonym.rbfox1]="RBFOX1"
rownames(counts.seqc)[rownames(counts.seqc)==synonym.cavin1]="CAVIN1"
rownames(counts.seqc)[rownames(counts.seqc)==synonym.cytor]="CYTOR"
rownames(counts.seqc)[rownames(counts.seqc)==synonym.eva1b]="EVA1B"


################################################################################
# import and incorporate metadata
################################################################################

metadata.seqc.original0=((fread(file.path(params$datapath, "bulkrna_patients/seqc/GSE49711_series_matrix.txt"), skip=65, fill=T, header=F) %>% t) %>% as.data.frame %>% mutate(V1=clean.name(V1))) %>% col2names(., "V1") 
metacols=make.names(metadata.seqc.original0[1,])
metadata.seqc.original= metadata.seqc.original0 %>% remove.first.row

seqc.fields= (metadata.seqc.original %>% apply(., c(1,2), get.seqc.field))[1,] %>% as.character

metadata.seqc.original=metadata.seqc.original %>% apply(., c(1,2), format.seqc.entry) 

metacols2=metacols
metacols2[metacols=="X.Sample_characteristics_ch1"]=seqc.fields[metacols=="X.Sample_characteristics_ch1"] %>% make.names
colnames(metadata.seqc.original)=metacols2
## getting rid of extra duplicated columns from protocol etc
metadata.seqc.original=metadata.seqc.original[,!duplicated(metacols2)]
metadata.seqc.original=metadata.seqc.original %>% as.data.frame %>% names2col(., "sample.id") %>% mutate( high.risk.original=high.risk, inss_stage=inss.stage, MYCN_status=mycn.status) 

seqc.clinical=fread(file.path(params$datapath, "bulkrna_patients/seqc/clinicalvars.csv")) %>% as.data.frame %>% mutate(NB.ID=get("Patient ID")) %>% col2names(., coll="NB.ID")

#import metadata with survival from seqc
theader0=(fread(file.path(params$datapath, "bulkrna_patients/seqc/GSE62564_series_matrix.txt"), skip=50, header=F, fill=T) %>% t)[1, ] %>% make.names

seqc=(fread(file.path(params$datapath, "bulkrna_patients/seqc/GSE62564_series_matrix.txt"), skip=50, fill=T)[, 2:499] %>% t) 
seqc.formatted=seqc %>% apply(., c(1,2), format.seqc.entry) 
seqc.fields2= (seqc %>% apply(., c(1,2), get.seqc.field))[1,] %>% as.character
theader=theader0[2:length(theader0)]
theader2=theader
theader2[theader=="X.Sample_characteristics_ch1"]=seqc.fields2[theader=="X.Sample_characteristics_ch1"]
seqc.formatted=as.data.frame(seqc.formatted)
colnames(seqc.formatted)=theader2 %>% make.names 
rownames(seqc.formatted)= clean.name(rownames(seqc.formatted))
metadata.seqc.2=seqc.formatted %>% names2col(., coln="NB.ID") %>% mutate(Overall_Survival_Time_in_Days=os.day, Event_Free_Survival_Time_in_Days=efs.day) %>% left_join(., seqc.clinical,  by="NB.ID") %>% mutate(MYCN_status=get("MYCN status"), inss_stage=get("INSS Stage"), days_to_birth=-Age, sample.id=NB.ID) %>% col2names(., "sample.id")

################################################################################
# merge both metadata tables
################################################################################

mtso.reduced=metadata.seqc.original[, c("sample.id", "dataset", "age.at.diagnosis", "mycn.status", 
"inss.stage", "class.label", "progression", "death.from.disease", 
"high.risk.original", "MYCN_status")]

mts2.reduced=metadata.seqc.2[, c("sample.id", "training.validation", "age", "efs.day", "efs.bin", "os.day", 
"os.bin", "a_efs_all", "b_os_all", "c_sex_all", "d_fav_all", 
"e_efs_hr", "f_os_hr", "Overall_Survival_Time_in_Days", 
"Event_Free_Survival_Time_in_Days", "Patient ID", "Country", 
"inss_stage", "Age", "HighRisk", "Sex_Imputed", 
"OS_bin", "days_to_birth")]


metadata.seqc.full=mtso.reduced %>% as.data.frame %>% left_join(.,
mts2.reduced , by="sample.id") %>% mutate(inss_stage=paste0("Stage ",inss_stage))%>% mutate(inss_stage=edit.stage23(inss_stage), high.risk=high.risk.original, Vital_Status=format.status(death.from.disease), Event_Free_Survival=format.status(efs.bin)
                                                                                            ) %>% col2names(., "sample.id")

################################################################################
# seqc dataset overview and filtering to stage
################################################################################

seqc.before.filtering=transferplot(so=NULL, df=metadata.seqc.full, labelvars = c( "inss_stage", "MYCN_status", "high.risk.original") , colorlist=allcolors)+ggtitle("Overview of entire SEQC dataset")
seqc.before.filtering 

mts=metadata.seqc.full

mts= mts %>% mutate(MYCN_status=lapply(metadata.seqc.full %>% pull(MYCN_status), function(x) generalreplace(x, c("0", "1", "N/A"), c("Not Amplified", "Amplified", "Unknown")))  %>% unlist)

seqc.after.filtering=transferplot(so=NULL, df=mts, labelvars = c( "inss_stage", "MYCN_status") , colorlist=allcolors)+ggtitle("Overview of entire dataset")

seqc.after.filtering

```

load CCRI dataset

```{r loadccri}

################################################################################
# Data is separated into MYCN and non MYCN samplew shich must be merged first
################################################################################
metadata.stm.mycn=fread(file.path(params$datapath, "bulkrna_patients/ccri/metadata_MYCN.csv"))
metadata.stm.nonmycn=fread(file.path(params$datapath, "bulkrna_patients/ccri/metadata_nonMYCN.csv"))      
metadata.stm.clinical0=fread(file.path(params$datapath, "bulkrna_patients/ccri/clinical_eb_up_v2.csv"), skip=1)  %>% as.data.frame %>% col2names(., "omics_id")

metadata.stm.clinical=metadata.stm.clinical0[, colnames(metadata.stm.clinical0)[!duplicated(colnames(metadata.stm.clinical0))]]


all.cols=union(colnames(metadata.stm.mycn), colnames(metadata.stm.nonmycn))
shared.cols=intersect(colnames(metadata.stm.mycn), colnames(metadata.stm.nonmycn))

metadata.stm.full=rbind(metadata.stm.mycn %>% dplyr::select(all_of(shared.cols)), metadata.stm.nonmycn %>% dplyr::select(all_of(shared.cols))) %>% as.data.frame %>% col2names(., "omics_id")

################################################################################
# further formatting of metadata
################################################################################

edit.mycn=Vectorize(function(x) if(x=="YES"){return("Amplified")}else{return("Not Amplified")}, USE.NAMES=F)
edit.stage=Vectorize(function(x) if(is.na(x) || x==""){return(NA)}else{return(paste0("Stage ", x))}, USE.NAMES=F)
edit.17q=Vectorize(function(x){
  if(!is.na(x)){
  if(x=="YES"|| x=="yes"){return("yes")}else{return("no")}}else{return(NA)}}, USE.NAMES=F)

metadata.stm.full.clinical=metadata.stm.full %>% col2names(., "omics_id") %>% mutate(MYCN_status=edit.mycn(mna),
x17q_gain=edit.17q(x17q_gain),                                                                                     inss_stage=edit.stage23(edit.stage(metadata.stm.clinical[metadata.stm.full %>% rownames, "INSSStadium" ])),
                                                                                      Event_Free_Survival_Time_in_Days=metadata.stm.clinical[metadata.stm.full %>% rownames, "EFS_dur" ] ,
                                                                                     time_to_death= metadata.stm.clinical[metadata.stm.full %>% rownames, "tod_dur" ] , Event_Free_Survival=metadata.stm.clinical[metadata.stm.full %>% rownames, "EFS" ], 

Vital_Status=metadata.stm.clinical[metadata.stm.full %>% rownames, "Tod" ],
                                  date_of_birth=metadata.stm.clinical[metadata.stm.full %>% rownames, "GEBDAT" ]

,date_of_diagnosis=metadata.stm.clinical[metadata.stm.full %>% rownames, "DXDAT" ],
date_of_death=metadata.stm.clinical[metadata.stm.full %>% rownames, "death_date" ],
Event_type=metadata.stm.clinical[metadata.stm.full %>% rownames, "Event_type" ],
MYCN_status2=metadata.stm.clinical[metadata.stm.full %>% rownames, "NMCY" ])

################################################################################
# load counts from CCRI
################################################################################

counts.stm=fread(file.path(params$inputdata, "bulkrna_patients/stm/full_omics_expression_data_normalized_counts.csv")) %>% as.data.frame
counts.d.stm=counts.stm[!duplicated(counts.stm %>% pull(gene_name)), ] %>% col2names(., "gene_name")
stm.intersect=intersect(as.data.frame(counts.stm) %>% colnames, metadata.stm.full %>% pull(omics_id))
counts.stm.ready=counts.d.stm %>% select(all_of(stm.intersect))

dds.stm=readRDS(file.path(params$datapath, "bulkrna_patients/stm/full_omics_dds.rds"))

#has no appropriate gene names
counts.stm.dds.raw=assay(dds.stm)

dnms=rowData(dds.stm) %>% as.data.frame %>% pull(gene_name) 

counts.stm.dds.formatted=counts.stm.dds.raw[!duplicated(dnms), ] %>% giverownames(., dnms[!duplicated(dnms)])

stm.intersect.dds=intersect(as.data.frame(counts.stm.dds.formatted) %>% colnames, metadata.stm.full.clinical %>% pull(omics_id))

counts.stm.dds.ready=counts.stm.dds.formatted[, stm.intersect.dds]

############################################################################
# finding synonyms for marker genes and harmonising them across datasets
############################################################################

synonym.h2afj=c( "H2AJ", "FLJ10903","MGC921")[c( "H2AJ", "FLJ10903","MGC921") %in% rownames(counts.stm.dds.ready)]

rownames(counts.stm.dds.ready)[rownames(counts.stm.dds.ready)==synonym.h2afj]="H2AFJ"

################################################################################
#filter experiments to remove problematic situations, for example keep only diagnosis
################################################################################

stm.before.filtering=transferplot(so=NULL, df=metadata.stm.full.clinical[stm.intersect,], labelvars = c( "inss_stage", "MYCN_status", "material_general", "tp") , colorlist=allcolors)+ggtitle("Overview of entire CCRI dataset")
stm.before.filtering 

mtc=metadata.stm.full.clinical[ stm.intersect.dds,] %>% filter(tp=="DE")

### complement metadata with the columns in the bicu metadata
additional.columns=setdiff(colData(dds.stm) %>% as.data.frame %>% colnames, colnames(mtc))

mtc.complete0= cbind(mtc, (colData(dds.stm) %>% as.data.frame)[rownames(mtc), additional.columns ])

fixed.efstime=lapply(1:nrow(mtc.complete), function(num) {
ptid= mtc.complete[num, "patient_o_id"]
survivaltimes=mtc.complete %>% filter(patient_o_id==!!ptid) %>% pull(Event_Free_Survival_Time_in_Days)
val=survivaltimes[!is.na(survivaltimes)] %>% unique
if(length(val)<1){
 return(NA) 
}else{
 return(val) 
}
}) %>%Reduce(c, .)

fixed.efs=lapply(1:nrow(mtc.complete), function(num) {
ptid= mtc.complete[num, "patient_o_id"]
survivals=mtc.complete %>% filter(patient_o_id==!!ptid) %>% pull(Event_Free_Survival)
val=survivals[!is.na(survivals)] %>% unique
if(length(val)==0){
 return(NA) 
}else{
 return(val) 
}
}) %>%  Reduce(c, .)


pts= mtc.complete %>% group_by(patient_o_id) %>% summarise(counts=n()) %>% as.data.frame %>% filter(counts<2) %>% pull(patient_o_id)
mtc.complete=mtc.complete0 %>% mutate(Event_Free_Survival=fixed.efs, Event_Free_Survival_Time_in_Days=fixed.efstime) %>% filter(patient_o_id %in% pts)


stm.after.filtering=transferplot(so=NULL, df=mtc.complete, labelvars =c( "inss_stage", "MYCN_status", "material_general", "tp") , colorlist=allcolors)+ggtitle("Overview of filtered dataset")

stm.after.filtering

```

SEQC. preliminary analysis

```{r}

################################################################################
# handling data via DESeq2 but not normalising as the data has been already processed in seveal ways
################################################################################

dds.seqc <- DESeqDataSetFromMatrix(apply(counts.seqc[, rownames(mts)], c(1,2), function(x) (2**x) %>% as.integer), colData= mts, design = ~1)


vsd.seqc=assay(vst(dds.seqc))

################################################################################
# performing PCA on the VST counts
################################################################################

gpca=prcomp(vsd.seqc %>% t)

gpcdf.seqc=gpca$x %>% as.data.frame %>% names2col(., "File.ID")

################################################################################
# incorporating metadata to the pca data frame for ease of plotting
################################################################################
metavars=colnames(mts)
for(mv in metavars){
gpcdf.seqc[, mv]= mts[gpcdf.seqc %>% pull("File.ID"), mv] %>% unname # incorporating 
}
comp1=1
comp2=2
```

CCRI preliminary analysis.

```{r}

dds.bicu <- DESeq(DESeqDataSetFromMatrix(counts.stm.dds.ready[, rownames(mtc.complete)], colData=mtc.complete , design = ~library_layout+material))

vsd.ccri=assay(vst(dds.bicu), normalised=T)

################################################################################
# performing PCA on the VST counts
################################################################################

gpca=prcomp(vsd.ccri %>% t)
gpcdf.ccri=gpca$x %>% as.data.frame %>% names2col(., "File.ID")

################################################################################
# incorporating metadata to the pca data frame for ease of plotting
################################################################################
metavars=colnames(mtc.complete)
for(mv in metavars){
gpcdf.ccri[, mv]= mtc.complete[gpcdf.ccri%>% pull("File.ID"), mv] %>% unname # incorporating 
}
comp1=1
comp2=2

################################################################################
# plotting the PCA
################################################################################
colorbys=c("library_layout", "analysis", "material", "MYCN_status", "inss_stage")#"MYCN_status"
#"tissue_or_organ_of_origin"
pcas.ccri=lapply(colorbys, function(xx){
ggplot(gpcdf.ccri, aes(x=PC1, y=PC2, color=!!sym(xx)))+geom_point( size=sz)+theme_classic()
}) %>% Reduce('+', .)+plot_layout(ncol=3, nrow=3)

sca=13
rr=600
sz=2
library(ggrepel)
sca=15
tpng(path=params$plotpath, paste_("pcas_bulkrna_vst_ccri"),  width=w*sca*4/3, height=w*sca, res=rr)
pcas.ccri
dev.off()

```

CCRI preliminary analysis. (only TUM)

```{r}

mtc.tum=mtc.complete %>% filter(material_general=="TUM")

dds.bicu <- DESeq(DESeqDataSetFromMatrix(counts.stm.dds.ready[, rownames(mtc.tum)], colData=mtc.tum , design = ~library_layout+material))

vsd.ccri.tum=assay(vst(dds.bicu), normalised=T)

################################################################################
# performing PCA on the VST counts
################################################################################


gpca=prcomp(vsd.ccri %>% t)
gpcdf.ccri=gpca$x %>% as.data.frame %>% names2col(., "File.ID")

################################################################################
# incorporating metadata to the pca data frame for ease of plotting
################################################################################
metavars=colnames(mtc.complete)
for(mv in metavars){
gpcdf.ccri[, mv]= mtc.complete[gpcdf.ccri%>% pull("File.ID"), mv] %>% unname # incorporating 
}
comp1=1
comp2=2

################################################################################
# plotting the PCA
################################################################################
colorbys=c("library_layout", "analysis", "material", "MYCN_status", "inss_stage")#"MYCN_status"
#"tissue_or_organ_of_origin"
pcas.ccri=lapply(colorbys, function(xx){
ggplot(gpcdf.ccri, aes(x=PC1, y=PC2, color=!!sym(xx)))+geom_point( size=sz)+theme_classic()
}) %>% Reduce('+', .)+plot_layout(ncol=3, nrow=3)


sca=13
rr=600
sz=2
library(ggrepel)
sca=15
tpng(path=params$plotpath, paste_("pcas_bulkrna_vst_ccri"),  width=w*sca*4/3, height=w*sca, res=rr)
pcas.ccri
dev.off()

```

CCRI unfiltered dataset

```{r}

dds.bicu.unfiltered <- DESeq(DESeqDataSetFromMatrix(counts.stm.dds.ready[, rownames(metadata.stm.full.clinical)], colData=metadata.stm.full.clinical , design=~1))

vsd.ccri.unfiltered=assay(vst(dds.bicu.unfiltered), normalised=T)

################################################################################
# performing PCA on the VST counts
################################################################################

gpca=prcomp(vsd.ccri.unfiltered %>% t)
gpcdf.ccri.unfiltered=gpca$x %>% as.data.frame %>% names2col(., "File.ID")

################################################################################
# incorporating metadata to the pca data frame for ease of plotting
################################################################################
metavars=colnames(metadata.stm.full.clinical)
for(mv in metavars){
gpcdf.ccri.unfiltered[, mv]= metadata.stm.full.clinical[gpcdf.ccri.unfiltered%>% pull("File.ID"), mv] %>% unname # incorporating 
}
comp1=1
comp2=2

################################################################################
# plotting the PCA
################################################################################
colorbys=c("analysis", "material", "MYCN_status", "inss_stage")#"MYCN_status"
#"tissue_or_organ_of_origin"
pcas.ccri=lapply(colorbys, function(xx){
ggplot(gpcdf.ccri.unfiltered, aes(x=PC1, y=PC2, color=!!sym(xx)))+geom_point( size=sz)+theme_classic()
}) %>% Reduce('+', .)+plot_layout(ncol=3, nrow=3)

sca=13
rr=600
sz=2
library(ggrepel)
sca=15
tpng(path=params$plotpath, paste_("pcas_bulkrna_vst_ccri"),  width=w*sca*4/3, height=w*sca, res=rr)
pcas.ccri
dev.off()

```

TARGET Preliminary analysis: load DEseq object, estimate size factors and obtain VST for count matrix.

```{r}
minexpts=3  ##minimum number of experiments in which the gene must be present
minreads=1 # minimum number of reads per experiment
nvargenes=200

################################################################################
# normalising depth using DEseq and obtaining variance stabilising transformation
################################################################################

dds.target <- DESeq(DESeqDataSetFromMatrix(counts.target[, rownames(mt)], colData= mt, design = ~1))
vsd.target=assay(vst(dds.target))

################################################################################
# perfomring PCA on the VST counts
################################################################################

gpca.target=prcomp(vsd.target %>% t)

gpcdf.target=gpca.target$x %>% as.data.frame %>% names2col(., "File.ID")

################################################################################
# incorporating metadata to the pca data frame for ease of plotting
################################################################################
metavars=colnames(mt)
for(mv in metavars){
gpcdf.target[, mv]= mt[gpcdf.target %>% pull("File.ID"), mv] %>% unname # incorporating 
}
comp1=1
comp2=2

################################################################################
# plotting the PCA
################################################################################
colorby="MYCN_status"
#"tissue_or_organ_of_origin"
sca=13
rr=600
sz=2
library(ggrepel)
sca=15
tpng(path=params$plotpath, paste_("pca_bulkrna_vst"),  width=w*sca*4/3, height=w*sca, res=rr)
ggplot(gpcdf.target, aes(x=PC1, y=PC2, color=!!sym(colorby)))+geom_point( size=sz)+theme_classic()
dev.off()
```

Calculating signature scores for the patients, ranking by percentile , and plotting. Supplementary figure 7 d

```{r signatures}

## relevant signatures

paired.sigs=rbind(c("C13", "C5"), c("M5","M8"))
    
#different lower and upper percentages of the signatures
l1=0.33
l2=0.45
h1=0.67
h2=0.55

### parameters for plots to be generated

parlist=list(
list(ds0="seqc", vsd=vsd.seqc, gpcdf=gpcdf.seqc, filterstage=T, lowthresh=l1, topthresh=h1, plot.all.pairs=T),
list(ds0="seqc", vsd=vsd.seqc, gpcdf=gpcdf.seqc, filterstage=F, lowthresh=l1, topthresh=h1, plot.all.pairs=T),

)

#looping through parameter sets to generate plots
lapply(1:length(parlist), function(pp){

  
ds0=parlist[[pp]]$ds0
vsd=parlist[[pp]]$vsd
gpcdf=parlist[[pp]]$gpcdf
filterstage=parlist[[pp]]$filterstage

plot.facets=F
calculate.signatures=T
plot.all.pairs=parlist[[pp]]$plot.all.pairs

#legend size
ls=8
## point size for the signature combo plots
sz=1
if(ds0=="target"){
gpcdf=gpcdf %>% mutate(had.bm.relapse= grepl("Bone Marrow", gpcdf$Site.of.Relapse)) 
}

lowthresh=parlist[[pp]]$lowthresh
topthresh=parlist[[pp]]$topthresh
sca.pairs=.6

################################################################################
# plotting the PCA
################################################################################
colorby="MYCN_status"
#"tissue_or_organ_of_origin"
sca=13
rr=600
sz=2
library(ggrepel)
sca=15
tpng(path=params$plotpath, paste_("pca_seqc_bulkrna_", ds, "vst", colorby),  width=w*sca*4/3, height=w*sca, res=rr)
ggplot(gpcdf, aes(x=PC1, y=PC2, color=!!sym(colorby)))+geom_point( size=sz)+theme_classic()
dev.off()

if(calculate.signatures){
#################################################################################
# calculating signatures
################################################################################

gsva_output=GSVA::gsva(vsd %>% as.matrix(),
                         append(markers.wt.shared, markers.mut.shared ),
                         method = "ssgsea")
signature.scores=gsva_output %>% t 

sigs=signature.scores %>% colnames
for(mv in sigs){
gpcdf[, mv]= signature.scores[gpcdf %>% pull("File.ID"), mv] %>% unname # incorporating 
}

calculate.signatures=F
}

signature.vars=signature.scores %>% colnames

#########
# apply any filters
#########

if(filterstage){
  gpcdf= gpcdf %>% filter(inss_stage=="Stage 4")
ds=paste0(ds0, "stage4")
}else{
  ds=ds0
}

fwrite(gpcdf %>% dplyr::select(!starts_with("PC")), file.path(params$datapath, "bulkrna_patients", paste0("survival_dataset_", ds, ".csv") )  )

################################################################################
# plotting signature score point plots for C13 and C5. Figure 7 d
################################################################################

if(plot.all.pairs){

lapply(1:nrow(paired.sigs), function(xx){
sig1=paired.sigs[xx, 1]
sig2=paired.sigs[xx, 2]  
  fcat(ds, sig1, "vs", sig2, "...")

sca=13
rr=600

library(ggrepel)

#day
sigs.mycn=(ggplot(gpcdf, aes(x=!!sym(sig1), y=!!sym(sig2), color=MYCN_status))+geom_point( size=sz)+theme_classic()+scale_color_manual(values=allcolors[["MYCN_status"]])+theme(legend.position = "bottom")+ggtitle(ds)) %>% ggMarginal(., type="density", groupFill=T, groupColour=T)

sigs.stage=(ggplot(gpcdf, aes(x=!!sym(sig1), y=!!sym(sig2), color=inss_stage))+geom_point( size=sz)+theme_classic()+scale_color_manual(values=allcolors[["inss_stage"]])+theme(legend.position = "bottom")+guides(color = guide_legend(nrow = 1))+ggtitle(ds) )%>% ggMarginal(., type="density", groupFill=T, groupColour=T)

tpdf(path=params$plotpath, paste_("scatterplot_",ds, "vst", sig1, "vs", sig2, "_bulkrna_"),  width=pw*sca.pairs*2, height=pw*sca.pairs)

print(grid.arrange(sigs.mycn,sigs.stage, ncol=2))
dev.off()

})
}
  
## functions to calculate percentile based actions, among them, survival analyses.
pctile= function(varname, val){
 vals=gpcdf %>% pull(!!sym(varname))
  dd= ecdf(vals) 
  v=dd(val)
  
  v
}

classify= Vectorize(function(varname, val){
  vals=gpcdf %>% pull(!!sym(varname))
 dd= ecdf(vals) 
  v=dd(val)
  
  if (v>=topthresh){return("high")}else{
  if (v<=lowthresh){return("low")}else{
  return("mid") }}
}, USE.NAMES=F)


format.censoring=Vectorize(function(x) if(x=="Censored"){return(0)}else{
  if(x %in% c("Event", "Progression", "Relapse", "Death")){return(1)}
  }, USE.NAMES=F)

add.vital.status=function(x){
 if(!("Vital_Status" %in% colnames(x))){
x= x %>% mutate(Vital_Status=format.status(death.from.disease))
 }
   x
}
})

```

## Session info etc.

Runtime: `r time_diff(SETUP_TIME)`

```{r}
sessionInfo()
```